sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/FileSourceStrategy.scala

```diff
@@ -189,7 +189,13 @@ object FileSourceStrategy extends Strategy with PredicateHelper with Logging {
       // Partition keys are not available in the statistics of the files.
       // `dataColumns` might have partition columns, we need to filter them out.
       val dataColumnsWithoutPartitionCols = dataColumns.filterNot(partitionSet.contains)
-      val dataFilters = normalizedFiltersWithoutSubqueries.flatMap { f =>
+      // Scalar subquery can be pushed down as data filter at runtime, since we always
+      // execute subquery first.
+      // It has no meaning to push down bloom filter, so skip it.
+      val normalizedFiltersWithScalarSubqueries = normalizedFilters
+        .filterNot(e => e.containsPattern(PLAN_EXPRESSION) && !e.containsPattern(SCALAR_SUBQUERY))
```

Comment:
thinking about this again. I think all non-correlated subqueries can be allowed? They will be evaluated before the actual plan execution.

Reply:
Actually, subquery expressions must be non-correlated when we reach here (they must have been rewritten into joins before). Then we need to deal with two kinds of subqueries:
1. foldable ones like scalar subquery. We can turn them into literals at the beginning and do filtering.
2. non-foldable ones like the DPP subquery. We will evaluate them when we have an initial partition list and need to filter it further.

Reply:
yes, and list subquery should also be rewritten to join. So we can only match scalar subquery here for data filter. DPP is a partition filter.

Reply:
it seems 2 and 3 are same but 2 is data filter and 3 is partition filter ?

Reply:
Can 2 be used in partition filters?

Reply:
Technically it can be. But there is a issue if we want to use data filte to help pruning partition directory https://github.com/apache/spark/pull/41088#discussion_r1281730562. Or, add one more field to do one more list partition directory with normalized data filter.

Reply:
I mean, something like `part_col = (select ... )`

Reply:
yes, it has been worked since https://github.com/apache/spark/pull/23802. So, both `partitionFilters` and `dataFilters` support contain subquery expression.

Reply:
ok, so we only need to covert scala subquery to literal because filter pushdown needs v1 filter which doesn't support subquery?

Reply:
yes, v1 pushdown filter seems only support literal.